



* Who I am
* Where I work
* What the lab does
    - How it's like a startup/small business
    - Ownership and leadership of ideas
    - Small team means able to take responsibilities and make decisions and
      then see them through implementation

* Microbiology is like debugging in production.  (eh, not in love with this analogy)
    - Build tools to ask specific questions (inspect state)
    - Combine those tools in various ways to ask bigger picture questions (observe behaviour)
    - ...
    - ...


* Explore "last mile" idea
    - Not big data
    - But needs a computer


* How I fit in
    - Database
    - Backend
    - Frontend
    - Systems and network admin


* Need for these things 
    * Viroverse
        - faceted searches
        - sample lineage
        - patient charts
    * TCozy
        - plots
        - experiment qc comparison
        - stickers
        - instrument setup files
        - layout guides
    * ISDB / HIRIS
        - top genes viz
    * ICE Floe
    * Redash
    * Galaxy
    * Methylation Station
    * A host of other tools usable from the web (https://indra.mullins.microbiol.washington.edu/)
    * Data analysis pipelines
        - Low-frequency drug-resistance mutations
        - HHV8
        - NuGEN data processing
    * internal tooling
        - Who's using Geneious?
            - this got me a free coffee!
        - recs
    * Source code: https://github.com/MullinsLab


* Technology
    * PostgreSQL
        - Views!
    * Perl, Python, JavaScript, Elm
    * Angular.js


* Kinds of work
    - Data management
        - Ingestion and storage
        - Retrieval in variety of formats
        - Querying
    - Analysis
    - Visualization


* Rewards
    - ...
    - these are _thinking_ roles, not just rote "run this script" roles
    - part of changing this figure (Leek & Peng 2015, see Evan's slides)
    - ...


* Thoughts from Evan (via slides/talk)
    - __Empathy__
    - Lab notebooks image: supplement these
    - “Designing software like this is about fitting into workflows which are
       not primarily computer-based, and which programmers do not get to define.
       Our job is to save labor, not add to it. To do that, we build software
       right up to the edges of what researchers are already doing.”
    - “Writing software to support lab work requires embracing and empathizing
       with the circumstances of lab work. We made a big improvement very
       recently just by providing printable guides for the arrangement of
       samples on PCR plates. Paired with the stickers we also helped people
       print out so they could stick them on these tiny tubes, we’re making this
       stuff go smoother.”


"Internal service core" for the lab
    > Service cores such as ours are doing research. Our centre is more a
    > partner than a routine service provider — the input affects the success
    > of the project and shapes broader questions. As Korf puts it, “Our
    > service isn't just doing. It's also part of the thinking.”
    https://www.nature.com/news/core-services-reward-bioinformaticians-1.17251
